libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/WitnessUniverse.hs

```diff
@@ -0,0 +1,915 @@
+  , NFData hashtype
+  , Eq hashtype
+  ) =>
+  HasWitness hashtype era
+  where
+  type ProofType hashtype era
+  type WitnessType hashtype era
+  type TypeHashed hashtype era
+  hash :: TypeHashed hashtype era -> hashtype
+  mkWitness :: ProofType hashtype era -> WitnessType hashtype era
+  getTypeHashed :: ProofType hashtype era -> TypeHashed hashtype era
+  prettyHash :: hashtype -> PDoc
+
+-- ============================================================
+
+-- | A WitBlock is designed to have five purposes
+--   1) To efficiently constrain objects to be witnessed when using constraint generators
+--      the (Set hashtype) allows efficient constraints like :: (member_ t) (lit (wbHash witblock))
+--   2) To efficiently compute witnesses from a 'hashtype'
+--      the (Map hashtype (ProofType hashtype era)) can be used like this
+--      case Map.lookup hash (wbMap witblock) of
+--        Just base -> mkWitness base
+--        Nothing -> error "Missing hash. perhaps generator did not constrain the hash to be witnessed?"
+--   3) When (HasWitness hashtype era) holds, the WitBlock can be computed from only [ProofType hashtype era]
+--      using 'getTypeHashed'. This makes Gen and CBOR instances, especially easy. We compute only with
+--      [ProofType hashtype era] and then reconstruct the rest
+--   4) WitBlock is a Monoid, so we can combine them easily
+--   5) We can easily make (Gen (WitBlock t era)), so we can make them for testing.
+data WitBlock t era where
+  WitBlock :: (Era era, HasWitness t era) => Set t -> Map t (ProofType t era) -> WitBlock t era
+
+instance ToExpr t => ToExpr (WitBlock t era) where
+  toExpr (WitBlock x y) = Tree.App "WitBlock" [toExpr x, toExpr y]
+
+wbHash :: WitBlock t era -> Set t
+wbHash (WitBlock x _) = x
+wbMap :: WitBlock t era -> Map t (ProofType t era)
+wbMap (WitBlock _ y) = y
+
+-- | when we print a WitBlock, we are only interested in the hashes, not the witnesses
+instance PrettyA (WitBlock t era) where
+  prettyA (WitBlock hashset _) = ppSet (prettyHash @t @era) hashset
+
+instance Show (WitBlock t era) where
+  show x = show (prettyA x)
+
+instance NFData (WitBlock t era) where
+  rnf (WitBlock x y) = deepseq (rnf x) (deepseq (rnf y) ())
+
+instance Eq (WitBlock t era) where
+  (WitBlock x y) == (WitBlock a b) = x == a && y == b
+
+-- ==========================================================================
+-- (HasWitness t era) Instances for several different types of hash and hashed types
+
+-- A short descriptive name for a long complicated thing
+type BodyHash era = SafeHash EraIndependentTxBody
+
+-- ========
+-- KeyHash and VKey
+-- KeyPair seems to be missing a lot of instances, so we define them here
+
+instance Typeable r => EncCBOR (KeyPair r) where
+  encCBOR (KeyPair x y) = encode $ Rec KeyPair !> To x !> To y
+
+deriving instance Typeable r => Eq (KeyPair r)
+
+instance ToExpr (KeyPair r) where
+  toExpr (KeyPair x y) = Tree.App "KeyPair" [toExpr x, Tree.App (take 10 (show y)) []]
```

Comment:
Instead of defining those instances here as orphans, could you please move those instances next to the definition of `KeyPair`

Reply:
done